Rate limit calls strictly based on time between calls to the API

dennypage opened the pull request:
Rate limit calls strictly based on time between calls to the API rather than a mix between call and completion times. This addresses the issue of an upper level app receiving updates at half the requested rate when periodic calls are made at the configured rate_limit.

This issue can easily be seen in Home Assistant, which sets the rate_limit to 60, and has a periodic timer that makes an async API call every 60 seconds to request updated values. Since each update takes 1-2 seconds each to complete, the result is that every other call is ignored and data from the previous call is returned instead.

NB: The caching of current time (variable now) is required to prevent a race condition with clock second turnover.

dennypage committed:
Rate limit calls strictly based on time between calls to the API rather than
a mix between call and completion times. This addresses the issue of an uppper
level app receiving updates at half the requested rate when periodic calls
are made at the configured rate_limit.

kbickar commented:
It doesn't look like the changes in sense_energy/asyncsenseable.py or sense_energy/senseable.py are doing anything

dennypage commented:
**asyncsenseable.py:**
Time() is a level 2 OS call, and retrieves the current time, which the OS stores in seconds and microseconds, truncated to the second. The caching of the current time is necessary because in the prior code the return value of time() may have changed between the two calls due to the clock second turning over.

Consider the following sequence of events, with a rate_limit of 60:

The first invocation of  update_realtime() occurs at time 1670098649:999999 (seconds:microseconds). The first call to time() will return a value of 1670098649. We spend a couple of microseconds doing various things in the interpreter, and the time is now 1670098650:000001. The second call to time() will return 1670098650, effectively 1 second later, which we will store in last_realtime_call.

The subsequent invocation of update_realtime() occurs exactly 60 seconds after the first invocation at 1670098709:999999. When we call time(), we will be given a value of 1670098709. To do the rate limit check we will compare this with the value of last_realtime_call (1670098650), and incorrectly determine that only 59 seconds have passed since the prior invocation. The result is that no data will be retrieved, and the value cached from the request 60 seconds ago will be returned instead.

Thread switching can dramatically increase the window between the calls to time(). Regardless, using the value from a single call to time() for both the comparison and assignment completely eliminates the issue.

**senseable.py:**
Same logic as the above description for asyncsenseable.py, but with the added note that senseable.py was not previously setting the value of last_realtime_call at all, but was dependent upon the setting in sense_api.py which was removed to address the issue of elapsed time during the REST call.

kbickar commented:
Instead of worrying about microseconds, it seems like the correct solution would be to have HA check every 61 seconds or set the rate limit to 59 seconds.

dennypage commented:
It's not a question of microseconds vs milliseconds vs seconds, etc. The window in fact could be more than a second on a loaded system using priority based scheduling. Attempting to have HA (and any other app that might use the library) code around the problem doesn't make sense, especially when it is trivial to fix in the library itself. 

In general, using any form of "the time offset should never be more that this" approach always leads to some form of pathologic failure situation. The "single point in time" approach is used specifically to prevent these complications. This style of coding is just a fact of life for when planning for multi-thread multi-processor systems.

If it matters, the new code is also quite a bit more efficient. Level 2 calls such as time() are very expensive in comparison to general code in the interpreter. Caching the result of the one call probably removes 40% or more of the total cost of that method.

kbickar commented:
But what I'm saying is you've cached the time call over a single machine instruction and removed single call to time replacing it with a variable allocation assignment and so while you might save a microsecond, it's not going to solve any problems you might be having with syncing.  There are going to be other issues.  Function calls aren't instant, the HA time scheduling system doesn't have nano-second accuracy, and HA isn't multithreaded.

It's clear that having HA set the rate limit to the exact same as the update rate is going to cause problems, so that should probably set to `ACTIVE_UPDATE_RATE - timeout`

dennypage commented:
FWIW, there are likely several dozen instructions between the determination of the current second between the two calls to time(). And yes, HA actually is multithreaded. Heavily multithreaded. As is the kernel. And even if everything, including HA, Python, and the kernel, were all single threaded on a single processor, the issue of clock rollover would still be present as that is driven by hardware.

The new code addresses all issues associated with having the update rate the same as the rate_limit. I've been running it in my HA instance for a while now.

kbickar commented:
HA does use some threads, but it mostly uses asyncio which is a single-threaded way of handling slower I/O events.  The change to sense_api.py is why it's not rate limiting.  As you guess there are lots of instructions that happen between the conditional and the assignment, imagine how many instructions there are in calling the function or the delay in the check of if the timer is ready.  There's always going to be a chance that a few more microseconds have passed unless a change is also made on the HA side to reduce the rate limit

dennypage commented:
Yes, the change in sense_api.py was an easy win and addresses the previously guaranteed 50% failure rate. But that change alone isn't sufficient to address all failure situations. Without the changes in asyncsenseable.py I expect that there is still around a 1 in 1000 chance of failure. And when the failure occurs, it persists for several cycles. I have seen this failure in testing--it's rare, but it does happen. Why not fix it? The use of a single call to time() is a trivial change and completely addresses the failure scenario. There is no down side to this.
